### PR TITLE
[Doc] Upgrade docs build system

### LINKED ISSUE #37944: [docs] [docs infra] Finish Sphinx migration
### Description

The Sphinx migration has been started but it needs to be completed. https://github.com/ray-project/ray/pull/35073
cc: @max

### Link

_No response_

### PR DESCRIPTION
**Pull request by peytondmurray**
## Why are these changes needed?

This PR upgrades the build system for the docs.

## Related issue number

Closes https://github.com/ray-project/ray/issues/37944.

Will contain all the changes in https://github.com/ray-project/ray/pull/39766, but will be reviewed in pieces via PRs against this branch.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by peytondmurray** [comment]
## Todo

- [ ] Move the parse_navbar_config etc. to e.g. custom_directives.py

**Comment by peytondmurray** [comment]
According to RTD support the build is timing out (??). Troubleshooting now.

**Comment by peytondmurray** [comment]
Note: I had to reduce the version of `watchfiles` to `0.19.0` due to the premerge check. I'd argue the constraints should not be applied to the doc build requirements, as they're separate from the rest of the ray requirements.

Once https://github.com/click-contrib/sphinx-click/pull/129, we can remove this dependency from the `requirements-doc.txt`.

**Comment by angelinalg** [comment]
- [ ] The Ask AI button doesn't have a clear border. It should also be black in non-dark-mode.
Screenshot 2023-11-20 at 10 08 39 AM

**Comment by emmyscode** [comment]
- [ ] Some inconsistencies on the main landing page with padding and a permanent light background for the code samples.
- [ ] How does the user select the Ray version?
- [ ] Footer includes Sphinx version information; do we want this?
- [ ] The AskAI button has a very faint background so it doesn't look like a clickable button; missing Ray logo
- [ ] In the rightmost column on pages, there's an "Edit on Github" button; is that something we want?

File: ci/ray_ci/doc_builder_container.py [hunk 1]
```diff
@@ -4,11 +4,13 @@
 class DocBuilderContainer(Container):
     def __init__(self) -> None:
         super().__init__("docbuild")
+        self.install_ray(build_type="doc")
 
     def run(self) -> None:
         self.run_script(
             [
                 "cd doc",
+                "pip install -r requirements-doc.txt",
```
Comment (can-anyscale) [inline ci/ray_ci/doc_builder_container.py:13]: you probably don't need this, since this is already been installed in doc.build.Dockerfile (the docbuild image in line 6)

File: ci/ray_ci/tests.env.Dockerfile [hunk 2]
```diff
@@ -18,13 +18,16 @@ RUN <<EOF
 set -euo pipefail
```
Comment (can-anyscale) [inline ci/ray_ci/tests.env.Dockerfile:19]: probably also don't need to install dashboard, so move the if condition and exit early here?

**Comment by can-anyscale** [comment]
nit: can we add a line here https://github.com/ray-project/ray/blob/master/ci/pipeline/determine_tests_to_run.py#L276 something like:

```
elif changed_file == ".readthedocs.yaml":
  RAY_CI_DOC_AFFECTED = 1
```

It will help to get CI results quicker. The entire test suite is run on this change despite this is only a doc change.

**Review by rkooo567 — APPROVED** [review]

File: ci/docker/doc.build.Dockerfile [hunk 3]
```diff
@@ -12,11 +12,4 @@ SHELL ["/bin/bash", "-ice"]
...
 RUN pip install -U --ignore-installed  \
-  -c python/requirements_compiled.txt \
-  -r python/requirements.txt \
-  -r python/requirements/test-requirements.txt \
-  -r python/requirements/lint-requirements.txt \
-  -r doc/requirements-doc.txt 
-
-RUN HOROVOD_WITH_GLOO=1 HOROVOD_WITHOUT_MPI=1 HOROVOD_WITHOUT_MXNET=1 \
-  pip install -U --ignore-installed  -c python/requirements_compiled.txt horovod
+  -r doc/requirements-doc.txt
```
Comment (can-anyscale) [inline ci/docker/doc.build.Dockerfile:15]: [1]

File: ci/ray_ci/doc_builder_container.py [hunk 4]
```diff
@@ -4,12 +4,13 @@
...
     def __init__(self) -> None:
         super().__init__("docbuild")
-        self.install_ray()
+        self.install_ray(build_type="doc")
 
     def run(self) -> None:
         self.run_script(
             [
                 "cd doc",
+                "pip install -r requirements-doc.txt",
```
Comment (can-anyscale) [inline ci/ray_ci/doc_builder_container.py:13]: this should have been installed in [1]?

File: ci/ray_ci/tests.env.Dockerfile [hunk 5]
```diff
@@ -17,25 +17,29 @@ RUN <<EOF
...
-elif [[ "$BUILD_TYPE" == "asan" ]]; then
-  pip install -v -e python/
-  bazel build $(./ci/run/bazel_export_options) --no//:jemalloc_flag //:ray_pkg
-elif [[ "$BUILD_TYPE" == "java" ]]; then
-  ./java/build-jar-multiplatform.sh linux
-  RAY_INSTALL_JAVA=1 pip install -v -e python/
-elif [[ "$BUILD_TYPE" == "clang" || "$BUILD_TYPE" == "asan-clang" || "$BUILD_TYPE" == "tsan-clang" ]]; then
-  ./ci/env/install-llvm-binaries.sh
-  pip install -v -e python/
+if [[ "$BUILD_TYPE" == "doc" ]]; then
```
Comment (can-anyscale) [inline ci/ray_ci/tests.env.Dockerfile:20]: nit, maybe:

```
if [[ "$BUILD_TYPE" == "doc" ]]; then
  exit 0
fi
```

to avoid indentation of the whole block after this

File: ci/pipeline/determine_tests_to_run.py [hunk 6]
```diff
@@ -273,6 +273,8 @@ def get_commit_range():
             ):
                 RAY_CI_DOCKER_AFFECTED = 1
                 RAY_CI_LINUX_WHEELS_AFFECTED = 1
+            elif changed_file == ".readthedocs.yaml":
```
Comment (can-anyscale) [inline ci/pipeline/determine_tests_to_run.py:276]: w00t look like it still builds everything because of ci/ci.sh; if you change the `ci/run` here https://github.com/ray-project/ray/blob/master/ci/pipeline/determine_tests_to_run.py#L339 to just `ci` it should builds only ci and doc

not related to this pr so up to you, if you find yourself continue to iterate, it'll save some ci time

**Review by can-anyscale — COMMENTED** [review]
mainly on the reinstallation of requirements-doc.txt

File: ci/ray_ci/doc_builder_container.py (hunk 4 above)
Comment (peytondmurray) [reply ci/ray_ci/doc_builder_container.py:13]: One would have thought so. But in actuality it seems like there were _old versions_ of sphinx and other dependencies that were being called when building the docs. IMO we should take a closer look at the CI system after we have this merged, because it's _really_ hard as someone who hasn't worked on the CI system to understand at the moment.

Comment (can-anyscale) [reply ci/ray_ci/doc_builder_container.py:13]: probably because doc.build.Dockerfile was installed with `-U --ignore-installed`?

**Review by jjyao — APPROVED** [review]
It's a big change. Let's merge after the 2.9 branch cut?

**Review by architkulkarni — APPROVED** [review]

File: doc/source/cluster/vms/getting-started.rst [hunk 7]
```diff
@@ -29,43 +29,50 @@ Setup
...
+         .. tab-item:: GCP
+            :sync: GCP
 
             .. code-block:: shell
 
                 $ pip install -U "ray[default]" google-api-python-client
 
-   .. tab:: Community Supported
+   .. tab-item:: Community Supported
+        :sync: Community Supported
```
Comment (architkulkarni) [inline doc/source/cluster/vms/getting-started.rst:54]: Screenshot 2023-11-21 at 10 57 42 AM
This part looks messed up. Can we revert the changes in this table?

Comment (architkulkarni) [reply doc/source/cluster/vms/getting-started.rst:54]: Specifically the fact that `:sync:` appears and the title of the tabs are duplicated in this table and similar tables on this page.

**Review by architkulkarni — CHANGES_REQUESTED** [review]

File: doc/source/cluster/vms/getting-started.rst (hunk 7 above)
Comment (peytondmurray) [reply doc/source/cluster/vms/getting-started.rst:54]: Nice find. `:sync:` should synchronize the tab state of all the tabs on this page; there's probably a spacing issue or similar causing this. Will fix.

**Review by aslonnie — COMMENTED** [review]
(will leave this for @can-anyscale to review; I trust him on this one.)

**Review by architkulkarni — APPROVED** [review]
Screenshot 2023-11-21 at 2 05 54 PM
Looks good now!

File: doc/source/train/images/logo.svg
Comment (matthewdeng) [inline doc/source/train/images/logo.svg:1]: @peytondmurray do you know what's going on here? It looks off in this view but seems to render as expected...

File: doc/source/tune/examples/index.rst [hunk 8]
```diff
@@ -4,6 +4,16 @@
...
 =================
 
+.. toctree::
+    :hidden:
+
+    ml-frameworks
+    experiment-tracking
+    hpo-frameworks
+    Other Examples <other-examples>
+    Exercises <exercises>
```
Comment (matthewdeng) [inline doc/source/tune/examples/index.rst:14]: I'm not sure why this is happening but it looks like the sub-toctrees are getting generated again at this level

image

**Review by bveeramani — APPROVED** [review]
Data changes LGTM

**Comment by angelinalg** [comment]
> It's a big change. Let's merge after the 2.9 branch cut?

Since branch cut is one week delayed, let's try to get this in 2.9. Thanks!

**Review by shrekris-anyscale — APPROVED** [review]
Approving for the Serve changes.

This is awesome, thanks so much for all your hard work! So many docs issues that I've run into over the last few months are all resolved. Super excited to see this released!

**Review by ArturNiederfahrenhorst — APPROVED** [review]

File: doc/source/train/images/logo.svg
Comment (peytondmurray) [reply doc/source/train/images/logo.svg:1]: Yes, I modified the SVG, removing some hardcoded coloring. This was done so that we can insert the logo directly into the page as a DOM element, and then use CSS styles to apply the correct colors to different parts of the SVG. The benefits of this approach are:

1. We are using the same CSS variable for all Ray branding (`var(--ray-blue)`). We can change this to whatever we want if it needs to be adjusted int he future
2. The text part of the logo is set from a CSS variable controlled by `pydata-sphinx-theme` (`var(--pst-color-text-base)`). When the theme gets switched, this color changes - meaning that the text part of the ray train logo switches color appropriately when the user switches themes.

You're right that from an outsider's view, it's pretty confusing to see this. Maybe I can modify the SVG so it looks a bit more natural?

File: doc/source/tune/examples/index.rst (hunk 8 above)
Comment (peytondmurray) [reply doc/source/tune/examples/index.rst:14]: Good find. I think this is happening because `tune/examples/index.rst` has a toctree with references to a bunch of pages. Further down, we _literally include those pages_. Each of these includes its own toctree as well, which gets concatenated onto the sidebar nav tree, which is built from the toctree. 

I'm really not sure why we are doing this. IMO we should change it; I'd vote to just remove the `include` directives, and make these pages completely separate. How would you like to resolve this? @matthewdeng

File: doc/source/train/images/logo.svg
Comment (matthewdeng) [reply doc/source/train/images/logo.svg:1]: Ohh yeah I personally have no familiarity with CSS so I had no idea that's what was happening 😆 

Naive question, but would it be possible to modify the SVG so it ends up with the entire "outline" of the logo, and CSS is used just for the coloring (as opposed to also drawing the circles and lines in the logo)?

File: doc/source/tune/examples/index.rst (hunk 8 above)
Comment (justinvyu) [reply doc/source/tune/examples/index.rst:14]: I think removing the includes in this `index.rst` is a good solution. Could we remove the toctree in `_toc.yaml` and create a `:toctree:` inside `tune/examples/index.rst` instead?

**Review by justinvyu — COMMENTED** [review]
Very cool!

One minor weird thing I notice is that the highlightning on diff codeblocks is broken sometimes (the last diff codeblock on these pages: [1](https://anyscale-ray--41115.com.readthedocs.build/en/41115/train/getting-started-pytorch.html#report-checkpoints-and-metrics) [2](https://anyscale-ray--41115.com.readthedocs.build/en/41115/train/getting-started-pytorch-lightning.html#prepare-your-lightning-trainer) [3](https://anyscale-ray--41115.com.readthedocs.build/en/41115/train/getting-started-transformers.html#report-checkpoints-and-metrics)). It's probably in other places too.

Screenshot 2023-11-22 at 12 56 57 PM

File: doc/source/rllib/index.rst [hunk 9]
```diff
@@ -54,7 +66,7 @@ PyTorch (or both, as shown below):
         <span data-ty="input">pip install "ray[rllib]" tensorflow torch</span>
     </div>
 
-.. margin::
+.. note::
```
[the remark this answers is not in the document]
Comment (peytondmurray) [reply doc/source/rllib/index.rst:69]: This is a problem with the underlying `rst`. This was an easy fix here, although it really isn't part of the upgrade as it affects the current docs as well.

image

**Comment by peytondmurray** [comment]
@justinvyu I don't see this on local builds:

**Comment by peytondmurray** [comment]
@justinvyu I took a look at the diff highlighting issue and it wasn't actually too difficult to update the style for diffs. I made the change to match github's colorblind (deuteranopia) setting:

image

File: doc/source/tune/examples/index.rst (hunk 8 above)
Comment (peytondmurray) [reply doc/source/tune/examples/index.rst:14]: > Could we remove the toctree in _toc.yaml and create a :toctree: inside tune/examples/index.rst instead?

Yes, these were part of the core changes made in this PR :tada: !

File: doc/source/train/images/logo.svg
Comment (peytondmurray) [reply doc/source/train/images/logo.svg:1]: Yep, I just checked that the CSS can override "inline" SVG styles, so I updated the SVG to do that. Now it looks more normal.

File: doc/source/tune/examples/index.rst [hunk 10]
```diff
@@ -1,24 +1,18 @@
...
 
 =================
 Ray Tune Examples
 =================
 
-.. tip:: Check out :ref:`the Tune User Guides <tune-guides>` To learn more about Tune's features in depth.
-
-.. _tune-recipes:
+.. toctree::
+    :hidden:
```
Comment (matthewdeng) [inline doc/source/tune/examples/index.rst:9]: Non-blocking, but how about we remove the `:hidden:` flag here? The page seems a little bare right now

image

**Review by matthewdeng — APPROVED** [review]
Thanks a ton 😄

**Review by can-anyscale — APPROVED** [review]

**Comment by JoshTanke** [comment]
@peytondmurray Was doc/source/ray-overview/eco-gallery.yml supposed to be deleted? Is there a new place to add integrations now?

**Comment by JoshTanke** [comment]
> @JoshTanke Yes, it was. Those integrations were being generated from that yaml file and inserted into `ray-overview/ray-libraries.rst`. Rather than have the extra step, just modify `ray-overview/ray-libraries.rst` directly.

makes sense, thank you for the info!
